### PR TITLE
health-server: implement rate-limiting

### PR DESCRIPTION
**Pull request by jshr-w**
This PR adds preliminary changes necessary to fully address #32820 & the corresponding CFP [design-cfps/#57](https://github.com/cilium/design-cfps/blob/main/cilium/CFP-32820-improve-scalability-of-health-probing.md).

Specifically, this PR:
- Changes the package being used for the ICMP pinger.
- Reconfigures the HTTP prober and ICMP pinger timeouts to 10s.
- Rate-limits the ICMP pinger and HTTP prober within the currently fixed probing interval.

Other changes to implement the CFP will be made in separate future PRs.

```release-note
connectivity health checking: improve the reliability of health checking at large scales by rate-limiting probes
```

### CODE REVIEW

File: pkg/health/server/prober.go [hunk 1]
```diff
@@ -283,45 +361,24 @@ func (p *prober) httpProbe(node string, ip string) *models.ConnectivityStatus {
...
-		for elem, primary := range node.Addresses() {
-			if _, addr := resolveIP(&node, elem, primary); addr != nil {
-				nodes[node.Name] = append(nodes[node.Name], addr)
-			}
-		}
-	}
-
-	return nodes
-}
+func (p *prober) runProbe() {
```
Comment (jshr-w) [inline pkg/health/server/prober.go:360]: This change broke the async prober (which I wanted to deprecate as a separate PR) 😅 

I wanted to have the function that sends out the probes block until all the results are written. But since this takes 60 +/- 10 seconds, the HTTP GET times out.

For now, I have the async prober return the result of the most recent probe, which seemed to make sense in the context of the new health probe behavior. (If this seems okay, I'll update the relevant doc). I'm not sure if it would be a better decision to have the async prober behave as it did before (without rate limiting) & deprecate that behavior from there.

**Comment by marseel** [comment]
Thanks for the PR @jshr-w !

I will review it early next week, knowing how health-checking is subtle I will need to pay extra attention :sweat_smile:

**Comment by jshr-w** [comment]
Sounds good to me, thanks @marseel! For sure this one has a lot of cases to think about 😅

File: pkg/health/server/prober.go [hunk 2]
```diff
@@ -346,39 +412,67 @@ func (p *prober) runHTTPProbe() {
...
 			}
 
-			resp := p.httpProbe(name, ip.String())
-			if resp.Status != "" {
-				scopedLog.Debugf("Failed to probe: %s", resp.Status)
+			ctx := context.Background()
+			httpWg.Add(1)
+			go p.httpProbe(name, ip.String(), ctx, httpResChan, &httpWg)
+			icmpWg.Add(1)
+			go p.icmpPing(name, ip.String(), ctx, icmpResChan, &icmpWg)
```
Comment (marseel) [inline pkg/health/server/prober.go]: I would probably simplify this a bit and move rate-limiting here. In case we have thousands of nodes, there is no point in starting thousands of goroutines, which would be waiting on rate limiter. So I would do something like this instead:
```
	if err := p.rateLimiter.Wait(ctx); err != nil {
		// Handle error
	} else {
               wg.Add(2)
               go p.httpProbe(name, ip.String(), ctx, httpResChan, &wg)
               go p.icmpPing(name, ip.String(), ctx, icmpResChan, &wg)
        }
```
And then:
```
	go func() {
		wg.Wait()
		close(icmpResChan)
                close(httpResChan)
	}()
```
Also, note that you would need to start goroutine with receive loops `for resp := range httpResChan {` and `for resp := range icmpResChan {` before starting `httpProbe` and `icmpPing` goroutines.

File: pkg/health/server/prober.go [hunk 3]
```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	pinger.Timeout = p.server.ProbeDeadline
+	pinger.Count = 1
+	pinger.OnFinish = func(stats *probing.Statistics) {
+		p.Lock()
+		defer p.Unlock()
```
Comment (marseel) [inline pkg/health/server/prober.go]: I don't think we need this lock anymore here? 
If we remove `icmpPingerRateLimiter` as well and pass `p.server.ProbeDeadline`, I believe we could change from having a method to a function: 
`func (p *prober) icmpPing(` -> `func icmpPing(`

File: pkg/health/server/prober.go [hunk 4]
```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+func (p *prober) httpProbe(node string, ip string, ctx context.Context, resChan chan<- connectivityResult, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-func (p *prober) httpProbe(node string, ip string) *models.ConnectivityStatus {
 	result := &models.ConnectivityStatus{}
+	if err := p.httpProbeRateLimiter.Wait(ctx); err != nil {
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
```
Comment (marseel) [inline pkg/health/server/prober.go]: Similar to `icmpPing`, if we move out `rateLimiter` and pass `p.server.Config.HTTPPathPort` through arguments, we could change it from method to function.

File: pkg/health/server/prober.go [hunk 5]
```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+		defer p.Unlock()
 
-	for _, status := range p.results {
-		status.Icmp = nil
+		if stats.PacketsRecv > 0 && len(stats.Rtts) > 0 {
+			if debugLogsEnabled {
+				scopedLog.WithField("rtt", stats.Rtts[0].Nanoseconds()).Debug("Ping successful")
+			}
+			result.Latency = stats.Rtts[0].Nanoseconds()
+			result.Status = ""
```
Comment (marseel) [inline pkg/health/server/prober.go]: nit: we don't need to clean `result.Status`

File: pkg/health/server/prober.go [hunk 6]
```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+		if debugLogsEnabled {
+			scopedLog.WithError(err).Debug("Failed to create pinger")
+		}
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	pinger.Timeout = p.server.ProbeDeadline
+	pinger.Count = 1
```
Comment (marseel) [inline pkg/health/server/prober.go:298]: note to myself: I see that we removed:
```
// FIXME: Doubling the default payload size to 16 is a workaround for GH-18177
	prober.Size = 2 * fastping.TimeSliceLength
```
The new implementation uses 24 bytes pings by default, so it should be fine. 
I will double-check that when experimentally manually though :)

**Review by marseel — COMMENTED** [review]
Thanks @jshr-w this looks great.
It's so much easier to understand the whole logic now. 
I left a few small nits to further simplify this part of the code. 
Once you address them, I will do a final pass of review and experiment manually with a new implementation to check it at a larger scale and also from correctness point of view.
Thanks!

File: pkg/health/server/prober.go (hunk 2 above)
Comment (jshr-w) [reply pkg/health/server/prober.go]: Ah this was bothering me but I didn't think of moving the wait here. Thanks for the suggestion, this is much better :)

File: pkg/health/server/prober.go (hunk 3 above)
Comment (jshr-w) [reply pkg/health/server/prober.go]: Good point, adjusted here and in httpProbe

File: pkg/health/server/prober.go [hunk 7]
```diff
@@ -234,34 +241,90 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+	if debugLogsEnabled {
+		scopedLog = log.WithFields(logrus.Fields{
+			logfields.NodeName: node,
+			logfields.IPAddr:   ip,
+		})
 	}
-}
 
-const httpPathDescription = "Via L3"
+	scopedLog.Debug("Pinging host")
```
Comment (marseel) [inline pkg/health/server/prober.go]: nit: this can be moved inside `if debugLogsEnabled {`

**Comment by marseel** [comment]
@jshr-w would you take a look at results of scale-test? 
It seems like we are still seeing unreachable endpoints / nodes:
image
before:
image

so it seems like with rate-limiting it's mostly better, but it still looks like some of the health-checks are failing. 
All of the data from your run can be found in `gs://cilium-scale-results/logs/scale-100-main/1730139724`
(this comes from run https://github.com/cilium/cilium/actions/runs/11559466550)

Most important links:
sysdump: gs://cilium-scale-results/logs/scale-100-main/1730139724/cilium-sysdump-final.zip
prometheus snapshot: gs://cilium-scale-results/logs/scale-100-main/1730139724/artifacts/prometheus_snapshot.tar.gz

**Comment by jshr-w** [comment]
With the initial code changes, I noticed a lot of ping timeouts reported by cilium-health. So, bumped the timeout to 10s to match with the HTTP probe. This seemed to remove all the timeouts reported in the sysdump.

Prometheus looks like this:
image

There were still some connectivity probes failing, towards the beginning of the CL2 run. Notes from investigation:
- That node with ~80 probes failing for ~1 minute was seen in basically every scale test run and was always a control plane node. Just noting in case this is relevant, it was odd to me that this was the case.
- The probes failing are both HTTP and ICMP. The error reported was generally a timeout.
- Correlated the timestamps between a probe timeout and the destination Cilium agent status -- this case was not frequent, but the Cilium health server API was not running yet, which I think explains why the probes failed.
- The more frequent case was that probes to a health endpoint timed out. My hypothesis is that perhaps the health endpoint was not up yet at the time of the probe.
- The probe errors seem transient; they eventually succeed after a ~3-minute span and continue to be successful over the rest of the CL2 test.

Added some logging to dig into the health endpoint status hypothesis.

File: pkg/health/server/prober.go (hunk 1 above)
Comment (marseel) [reply pkg/health/server/prober.go:360]: > This change broke the async prober (which I wanted to deprecate as a separate PR) 😅

You mean it broke synchronously prober, not async, right?
Even if that's the case, improvements in this PR outweigh it and it's a reasonable trade-off to deprecate synchronous prober. 
Alternatively, we could run a prober with higher qps to finish in shorter time-window always, for example < 30s so HTTP call finishes as well.

File: pkg/health/server/prober.go [hunk 8]
```diff
@@ -234,34 +241,90 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
...
+				scopedLog.WithField("rtt", stats.Rtts[0].Nanoseconds()).Debug("Ping successful")
+			}
+			result.Latency = stats.Rtts[0].Nanoseconds()
+		} else {
+			scopedLog.Debug("Ping failed")
+			result.Status = "Connection timed out"
+		}
+	}
+	pinger.SetPrivileged(true)
+	err = pinger.Run()
```
Comment (marseel) [inline pkg/health/server/prober.go:311]: nit: we could use `pinger.RunWithContext` with context that has timeout of `probeDeadline` to make sure it always finishes within timeout :)

**Review by marseel — APPROVED** [review]
Thanks for a great investigation!

> That node with ~80 probes failing for ~1 minute was seen in basically every scale test run and was always a control plane node. Just noting in case this is relevant, it was odd to me that this was the case.

It definitely makes sense as control-plane node is brought up first and then we add 100 nodes for workloads:
```
Sun, 03 Nov 2024 01:40:17 GMT
Run cilium/scale-tests-action/create-instance-group@d3ecfd83003f3e9c98ba125ca14933401d44918f
Sun, 03 Nov 2024 01:40:17 GMT
Run ./kops create ig workloads --subnet us-west1 \
```

So control-plane detects workload nodes quite fast and start probing them before they are even ready. So It's fine I think :)

I've also verified that memory/cpu usage was similar / slightly better. 

Before:
metric: Max Memory Usage: Perc99, value: 239.8417578125
metric: Max Memory Usage: Perc90, value: 235.01679687499998
metric: Max Memory Usage: Perc50, value: 229.904296875
metric: Average CPU Usage: Perc99, value: 0.07819446009389666
metric: Average CPU Usage: Perc90, value: 0.07238967136150234
metric: Average CPU Usage: Perc50, value: 0.06529577464788733

After:
metric: Max Memory Usage: Perc99, value: 237.0303515625
metric: Max Memory Usage: Perc90, value: 232.50546875
metric: Max Memory Usage: Perc50, value: 228.169921875
metric: Average CPU Usage: Perc99, value: 0.08231026128054343
metric: Average CPU Usage: Perc90, value: 0.07707037037037039
metric: Average CPU Usage: Perc50, value: 0.06894444444444445

Please update release-notes to something like:
```
connectivity health checking: improve the reliability of health checking at large scale by rate-limiting probes
```

Thanks for the amazing work :tada:

File: pkg/health/server/prober.go (hunk 1 above)
Comment (jshr-w) [reply pkg/health/server/prober.go:360]: Ah oops yes I meant the synchronous prober! 
Totally on the same page about deprecating the synchronous prober. Especially for larger clusters it would defeat the purpose of this change (and the probing interval change upcoming) => so then, to confirm:
- Currently, synchronous prober is returning the results of the most recent probe. I would need to update the docs to mention this behavior change?
- Maybe, as a separate PR, we would want to change the CLI connectivity tests to use `cilium-health status` instead of `cilium-health status --probe` and delete the ability to `--probe`? My impression was that --probe isn't recommended for use, so the deprecation process could be like this. But I'm not sure if it should be more formal, with a deprecation warning before removal like metrics?

Comment (marseel) [reply pkg/health/server/prober.go:360]: We can mark `--probe` option as hidden in case someone has some automation that uses it so we don't break them. 
We can also add update note saying that --probe is no longer supported and returns cached results instead.

Comment (jshr-w) [reply pkg/health/server/prober.go:360]: Got it, this makes sense! :) Thank you!

File: pkg/health/server/prober.go [hunk 9]
```diff
@@ -368,17 +425,15 @@ func (p *prober) runHTTPProbe() {
...
-	p.updateIcmpStatus()
-	return err
+	p.runICMPPinger()
 }
 
 // Stop disrupts the currently running RunLoop(). This may only be called after
 // a call to RunLoop().
 func (p *prober) Stop() {
-	p.Pinger.Stop()
+	//p.Pinger.Stop()
```
Comment (qmonnet) [inline pkg/health/server/prober.go]: Nit: Leftover to remove?

File: cilium-health/launch/launcher.go [hunk 10]
```diff
@@ -35,7 +35,7 @@ var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "cilium-health-la
 
 const (
 	serverProbeInterval  = 60 * time.Second
-	serverProbeDeadline  = 1 * time.Second
+	serverProbeDeadline  = 10 * time.Second
```
Comment (qmonnet) [inline cilium-health/launch/launcher.go:38]: This change seems unrelated to the rest of the commit. I'd recommend moving it to another commit maybe, and providing the motivation in the commit description.

File: Documentation/operations/upgrade.rst [hunk 11]
```diff
@@ -322,6 +322,8 @@ communicating via the proxy must reconnect to re-establish connections.
 * The Cilium cluster name validation cannot be bypassed anymore, both for the local and
   remote clusters. The cluster name is strictly enforced to consist of at most 32 lower
   case alphanumeric characters and '-', start and end with an alphanumeric character.
+* The option to run a synchronous probe using ``cilium-health status --probe`` is no longer supported,
+  and now returns the results of the most recent cached probe.
```
Comment (qmonnet) [inline Documentation/operations/upgrade.rst]:
```suggestion
  it is now a hidden option that returns the results of the most recent cached probe.
```

Is there any plan to remove the option in a future version? If so, please mention it in the note.

**Review by qmonnet — APPROVED** [review]
Thanks!

I didn't look at the code changes in details. Doc changes are trivial and look good; I have a few nits below, although not blocking.

File: Documentation/operations/upgrade.rst (hunk 11 above)
Comment (jshr-w) [reply Documentation/operations/upgrade.rst]: Thanks for the suggestion!

**Review by michi-covalent — APPROVED** [review]
🎏
